feat(k8s-plugin): add support for kubeconfig context switching

voyvodov opened the pull request:
Providing a new argument to choose between different kubeconfig contexts. 

<!--- Provide a general summary of your changes in the Title above -->


## Description
<!--- Describe your changes in detail -->
The idea behind is that user can be storing multiple clusters or environments in a single kubeconfig file and just switching context when wants to use specific one.

However, currently mayastor plugin doesn't support such change and it's always executed against `current-context`, which is inconvinient.

This PR is trying to add option/argument to allow user to specify a different context (as with kubectl) instead of the default one (`current-context`).

Depends on https://github.com/openebs/mayastor-control-plane/pull/1018

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
This is useful in case of single kubeconfig file with multiple contexts (e.g. clusters).


## Regression
<!-- Is this PR fixing a regression? (Yes / No) --> No

<!-- If Yes, optionally please include version or commit id or PR# that caused this regression, if you have these details. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Tested against several kind servers all in a single kubeconfig.
Tested with both provided and not provided context (e.g. use current one).

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.

voyvodov commented:
Sorry if code is not the best one - open to modify where and what is needed.

First time touching rust :)

tiagolobocastro reviewed (approved):
Also openebs/openebs might need changes as well, for the umbrella plugin which is built using this plugin as a subcommand

tiagolobocastro commented:
> Sorry if code is not the best one - open to modify where and what is needed.
> 
> First time touching rust :)

Wow first time, would not have guessed, you've done great! :crab:

voyvodov committed:
feat(k8s-plugin): add support for kubeconfig context switching

Providing a new argument to choose between different kubeconfig contexts.
This is useful in case of single kubeconfig file with multiple contexts (e.g. clusters)

Signed-off-by: Hristo Voyvodov <hristo.voyvodov@redsift.io>

voyvodov commented:
Will need another approval. Needed to fix merge conflicts on the submodules (didn't know it's auto-updated).

voyvodov commented:
> > Sorry if code is not the best one - open to modify where and what is needed.
> > First time touching rust :)
> 
> Wow first time, would not have guessed, you've done great! 🦀

I guess is easy if the codebase and tooling is good. If I need to start with blank probably will just discard :)

dsharma-dc commented:
helm workflow is failing is failing for dockerhub login. @pchandra19 ?

Abhinandan-Purkait commented on k8s/plugin/src/resources/mod.rs:
```diff
@@ -35,14 +35,19 @@ pub struct CliArgs {
     /// The kube client, with the correct install namespace.
     pub async fn client(&self) -> anyhow::Result<kube::Client> {
-        let mut config = kube_proxy::config_from_kubeconfig(self.kubeconfig.clone()).await?;
+        let opts = kube_proxy::kubeconfigoptions_from_context(self.context.clone());
```
```suggestion
        let opts = kube_proxy::kubeconfig_options_from_context(self.context.clone());
```

Continuing from the control-plane review

voyvodov replied:
Changed as suggested

tiagolobocastro commented:
> helm workflow is failing is failing for dockerhub login. @pchandra19 ?

I suspect it's because it's running on fork from external contributor, so it can't access the secrets.
I raised #740 to address this
Anyway I think we can probably try to merge since bors re-runs the same action

tiagolobocastro commented:
@voyvodov the control-plane PR merged, could you please update the submodule here?

voyvodov commented:
> @voyvodov the control-plane PR merged, could you please update the submodule here?

@tiagolobocastro updated.

tiagolobocastro commented:
@voyvodov  please update with rebase instead of Merge

voyvodov commented:
> @voyvodov please update with rebase instead of Merge

My bad...all ok now?

tiagolobocastro commented:
> > @voyvodov please update with rebase instead of Merge
> 
> My bad...all ok now?

Let's see if we can merge this before my other PR merges, as looks like there's a docker login issue on one of the workflows, but I think it's ok, let's try
bors merge

tiagolobocastro commented:
bors merge
